compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2491,9 +2491,26 @@ impl<'a> Parser<'a> {
         let (attrs, loop_block) = self.parse_inner_attrs_and_block()?;
 
         let kind = ExprKind::ForLoop(pat, expr, loop_block, opt_label);
+
+        self.recover_loop_else("for")?;
+
         Ok(self.mk_expr_with_attrs(lo.to(self.prev_token.span), kind, attrs))
     }
 
+    /// Recovers from an `else` clause after a loop (`for...else`, `while...else`)
+    fn recover_loop_else(&mut self, loop_kind: &'static str) -> PResult<'a, ()> {
+        if self.token.is_keyword(kw::Else) && self.may_recover() {
+            let else_span = self.token.span;
+            self.bump();
+            let else_clause = self.parse_else_expr()?;
+            self.sess.emit_err(errors::LoopElseNotSupported {
```

Comment:
One suggestion: It may be useful to label the loop/while/for keyword so we can say "else is attached to this loop". You should be able to pass it in.